### DISCUSSION TITLE
NoClassDefFoundError: com/pi4j/config/Config when running the java code om pi

### DISCUSSION THREAD

**Discussion by peterivarsson** [Pi4J/pi4j#171] [Q&A]
Hi,

Why does it complaining about NoClassDefFoundError
Is some dependencies missing?

/ Peter

****************** my java file
package se.peter.ivarsson.raspberry.gpio;
        
import com.pi4j.Pi4J;
import com.pi4j.context.Context;
import com.pi4j.io.i2c.I2C;
import com.pi4j.io.i2c.I2CConfig;
import com.pi4j.io.i2c.I2CProvider;

public class testGpio {

    private static final byte TCA9534_REG_ADDR_OUT_PORT = 0x01;
    private static final byte TCA9534_REG_ADDR_CFG = 0x03;

    public static void main(String[] args) throws InterruptedException {

        Context pi4j = Pi4J.newAutoContext();
        I2CProvider i2CProvider = pi4j.provider("linuxfs-i2c");
        I2CConfig i2cConfig = I2C.newConfigBuilder(pi4j).id("TCA9534").bus(1).device(0x70).build();

        try (I2C tca9534Dev = i2CProvider.create(i2cConfig)) {

			int config = tca9534Dev.readRegister(TCA9534_REG_ADDR_CFG);

            if (config < 0)
				throw new IllegalStateException(
						"Failed to read configuration from address 0x" + String.format("%02x", TCA9534_REG_ADDR_CFG));

			byte currentState = (byte) tca9534Dev.readRegister(TCA9534_REG_ADDR_OUT_PORT);

			if (config != 0x00) {
	
                System.out.println("TCA9534 is not configured as OUTPUT, setting register 0x" + String
						.format("%02x", TCA9534_REG_ADDR_CFG) + " to 0x00");

				tca9534Dev.writeRegister(TCA9534_REG_ADDR_CFG, (byte) 0x5a);
            }
        }
        
        pi4j.shutdown();
    }
}
**********************  pom.xml file
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <groupId>se.peter.ivarsson</groupId>
    <artifactId>raspberry-gpio-test</artifactId>
    <version>1.0.0-SNAPSHOT</version>
    <packaging>jar</packaging>
    <properties>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
        <maven.compiler.source>11</maven.compiler.source>
        <maven.compiler.target>11</maven.compiler.target>
        
        <slf4j.version>2.0.0-alpha5</slf4j.version>
        <pi4j.version>2.0</pi4j.version>
    </properties>

    <dependencies>
        <dependency>
            <groupId>org.slf4j</groupId>
            <artifactId>slf4j-api</artifactId>
            <version>${slf4j.version}</version>
        </dependency>
        <dependency>
            <groupId>org.slf4j</groupId>
            <artifactId>slf4j-simple</artifactId>
            <version>${slf4j.version}</version>
        </dependency>

        
        <dependency>
            <groupId>com.pi4j</groupId>
            <artifactId>pi4j-core</artifactId>
            <version>${pi4j.version}</version>
        </dependency>

        
        <dependency>
            <groupId>com.pi4j</groupId>
            <artifactId>pi4j-plugin-raspberrypi</artifactId>
            <version>${pi4j.version}</version>
        </dependency>
        <dependency>
            <groupId>com.pi4j</groupId>
            <artifactId>pi4j-plugin-pigpio</artifactId>
            <version>${pi4j.version}</version>
        </dependency>
        <dependency>
            <groupId>com.pi4j</groupId>
            <artifactId>pi4j-plugin-linuxfs</artifactId>
            <version>${pi4j.version}</version>
        </dependency>
    </dependencies>
   
    <build>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-jar-plugin</artifactId>
                <version>3.2.0</version>
                <configuration>
                    <archive>
                        <manifest>
                            <addClasspath>true</addClasspath>
                            <mainClass>se.peter.ivarsson.raspberry.gpio.testGpio</mainClass>
                        </manifest>
                    </archive>
                </configuration>
            </plugin>
        </plugins>
    </build>    
</project>

**Comment by eitch** [comment]
@peterivarsson On first glance all looks well. Do yo have the stack trace as well?

**Reply by peterivarsson** [reply]
This is the only trace I get.

pi@raspberrypi:~/pi-camera/bright_test $ !!
java -jar raspberry-gpio-test-1.0.0-SNAPSHOT.jar
Error: Unable to initialize main class se.peter.ivarsson.raspberry.gpio.testGpio
Caused by: java.lang.NoClassDefFoundError: com/pi4j/config/Config

**Comment by peterivarsson** [comment]
I have sript out everything except the row below.
It seems that the java code below doesn´t load the Pi4J dependecis

Context pi4j = Pi4J.newAutoContext();

I get this result:
ava -jar raspberry-gpio-test-1.0.0-SNAPSHOT.jar
Exception in thread "main" java.lang.NoClassDefFoundError: com/pi4j/Pi4J
        at se.peter.ivarsson.raspberry.gpio.testGpio.main(testGpio.java:29)
Caused by: java.lang.ClassNotFoundException: com.pi4j.Pi4J
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
        ... 1 more
**************************
/*
 * DEVELOPER:  Peter Ivarsson
 * PROJECT  :  Pi4J Test
 * FILENAME :  module-info.java
 *
 */

module se.peter.ivarsson.raspberry.gpio {

    requires com.pi4j;
}

**Comment by eitch** [comment]
Looking at your command, i think the pi4j library is missing in the classpath, which explains the ClassNotFoundException
